debian-bo-master/bo/source/net/fwtk-1.3/doc/man/netacl.8
.\" Marcus J. Ranum, 1993
.\" Copyright (C) Trusted Information Systems, Inc.
.\" All rights reserved
.TH NETACL 8 "23 August 1993"
.SH NAME
netacl - TCP network access control
.SH SYNOPSIS
.B netacl servicename
[invoked from inetd]
.SH DESCRIPTION
.IX  "netacl"  ""  "\(em network access control"
.LP
.B netacl
provides a degree of access control for TCP-based services
invoked from
.B inetd(8).
When a server is started,
.B inetd
invokes netacl with the name of the service requested, instead
of the actual server.
.B netacl
then searches its permissions information (read from
.B netperm-table
) to see if the host initiating the connection is authorized.
If the host is authorized, the real server process is invoked.
Otherwise,
.B netacl
exits. Acceptance or rejection of the service is logged via the
.B syslog
facility.
.LP
.B netacl
duplicates functionality found in other tools such as
.B log_tcp
by Wietse Venema, but is included with the toolkit
because it is a simpler implementation, contains no 
support for UDP services, and shares a common configuration
file with the rest of the toolkit components.
.SH OPTIONS
.LP
.B netacl
takes one parameter: the name of the service it is to
provide. This service name is appended to the string "netacl-"
to generate the name by which rules are read from the
.B netperm-table
configuration file. If invoked with no parameters,
the service is assumed to be the program name - in
case one needs to replace the executable of some daemon
with a copy of
.B netacl.
For example, if
.B netacl
is invoked as:
.br
.I netacl in.telnetd
.br
it will retrieve all the configuration rules for
.I netacl-in.telnetd.
.LP
The following configuration rules are recognized:
.TP
.B hosts [options]
specifies a host permission rule. Host permission rules are
in the form of:
.sp
.na
netacl-in.telnet permit-hosts host1 host2 -options
netacl-in.telnet deny-hosts host1 host2 -options
.ad
.sp
Following the
.B permit-hosts
or
.B deny-hosts
clause is a list of host names or IP-addresses, which can
contain wildcards. Host names are searched in order
until the first option (starting with a '-') is encountered,
at which point, if there is a match for that rule, it will
be accepted. If the rule is a deny-hosts rule, the program
will log the denial of service and exit. If it is a
permit-hosts rule, the options will be processed and
executed in order.
If there is no rule explicitly permitting or deny a service,
the service is denied. Options are:
.IP
.B \-exec executable [args]
specifies a program to invoke to handle the service. This
option
.B must
be the final option in the rule. A -exec
option
.B must
be present in every rule.
.IP
.B \-user userid
where userid is the numeric UID or the name from a login in
.B /etc/passwd which the program should be invoked as.
.IP
.B \-chroot rootdir
specifies a directory to which
.B netacl
should chroot(2) prior to invoking the service program. This
requires that the service program be present, and the pathname
for the executable be relative to the new root.
.SH EXAMPLES
.LP
In this example, the \\ line wraps have been added to fit
lines on the page. \\-escapes are not permitted in 
.B netperm-table,
they are here as part of the example only
.sp
.na
netacl-in.telnetd: permit-hosts 192.33.112.* \\
       -exec /usr/etc/in.telnetd
.br
netacl-in.ftpd: permit-hosts unknown \\
       -exec /bin/cat /usr/local/etc/noftp.txt
.br
netacl-in.ftpd: permit-hosts 192.33.112.* \\
       -exec /usr/etc/in.ftpd
.br
netacl-in.ftpd: permit-hosts * -chroot /home/ftp \\
       -exec /bin/ftpd -f
.ad
.LP
In the example above,
.B netacl
is configured to permit telnet only for hosts in a particular
subnet.
.B ftpd
is configured to accept all connections from systems that do
not have a valid DNS name ("unknown") and to invoke
.B cat
to display a file when a connection is made. This provides an
easy and flexible means of politely informing someone that they
are not permitted to use a service. Hosts in the specified
subnet are connected to the real FTP server in
.B /usr/etc/in.ftpd
but all connections from other networks are connected to a
version of the FTP server that is already chrooted to the
FTP area, effectively making all FTP activity "captive".
.SH INSTALLATION
.LP
To install
.B netacl
first place the executable in a system area, then modify
.B /etc/inetd.conf
as desired, replacing entries for the servers that will be
controlled via
.B netacl.
For example (syntax may differ slightly depending on O/S version)
the FTP service might be configured as:
.sp
.nf
ftp stream tcp nowait root /usr/local/etc/netacl in.ftpd
.fi
.sp
Once inetd.conf has been modified, restart or reload
.B inetd.
Verify installation by attempting a connection, and
monitoring the system logs.
.SH FILES
.PD 0
.TP 20
.B /etc/inetd.conf
.B netperm-table
.SH SEE ALSO
.BR netperm-table (5)
.BR inetd (8)
.SH BUGS


debian-bo-master/bo/source/net/fwtk-1.3/netacl/netacl.c
/*-
 * Copyright (c) 1993, Trusted Information Systems, Incorporated
 * All rights reserved.
 *
 * Redistribution and use are governed by the terms detailed in the
 * license document ("LICENSE") included with the toolkit.
 */

/*
 *	Author: Marcus J. Ranum, Trusted Information Systems, Inc.
 */
static	char	RcsId[] = "Header: netacl.c,v 1.5 94/11/01 11:56:09 mjr rel ";


#include	<syslog.h>
#include	<stdio.h>
#include	<sys/types.h>
#include	<sys/file.h>
#include	<pwd.h>
#include	<time.h>

#include	"firewall.h"

static	int	acceptrule();

static	char	rhost[512];
static	char	raddr[512];
static	char	*service;

main(ac,av)
int	ac;
char	*av[];
{
	Cfg	*cfp;
	Cfg	*cf;
	char	nbuf[512];

#ifndef	LOG_DAEMON
	openlog("netacl",LOG_PID);
#else
	openlog("netacl",LOG_PID|LOG_NDELAY,LFAC);
#endif

	if(ac == 1)
		service = av[0];
	else
		service = av[1];

	if(strlen(service) + 9 > sizeof(nbuf)) {
		syslog(LLEV,"fwtkcfgerr: name of av[](%s) too long",service);
		exit(1);
	}

	strcpy(nbuf,"netacl-");
	strcat(nbuf,service);

	if((cfp = cfg_read(nbuf)) == (Cfg *)-1) {
		syslog(LLEV,"fwtkcfgerr: %s exiting - cannot read configuration",service);
		exit(1);
	}

	if(peername(0,rhost,raddr,sizeof(rhost))) {
		syslog(LLEV,"fwtksyserr: %s exiting - cannot get peername",service);
		exit(1);
	}


	/* get list of permitted/denied hosts */
	for(cf = cfg_get("hosts",cfp); cf != (Cfg *)0; cf = cfg_get("hosts",(Cfg *)0)) {
		int	x;

		for(x = 0; x < cf->argc; x++) {

			/*
			if we hit a value with a leading '-' we are
			past processing the host names
			*/
			if(!strcmp(cf->argv[x],"-exec"))
				break;
			if(!strcmp(cf->argv[x],"-user"))
				break;
			if(!strcmp(cf->argv[x],"-chroot"))
				break;

			if(hostmatch(cf->argv[x],raddr))
				acceptrule(service,cf,cf->argc,cf->argv);
		}
	}

	syslog(LLEV,"deny host=%s/%s service=%s",rhost,raddr,service);
	exit(1);
}


static	int
acceptrule(srv,c,ac,av)
char	*srv;
Cfg	*c;
int	ac;
char	**av;
{
	int	avbase = 0;
	char	*chrdir = (char *)0;
	int	runuid = -1;
	int	rungid = -1;
	int	xecind = -1;
	int	x;

	if(c->flags & PERM_DENY) {
		syslog(LLEV,"deny host=%s/%s service=%s",rhost,raddr,srv);
		exit(1);
	}

	for(x = 0; x < ac; x++)
		if(!strcmp(av[x],"-exec")) {
			xecind = x;
			break;
		}

	if(xecind < 0) {
		syslog(LLEV,"fwtkcfgerr: no exec clause in rule line %d",c->ln);
		exit(1);
	}

	for(x = 0; x < xecind; x++)
		if(!strcmp(av[x],"-user")) {
			if((runuid  = mapuid(av[x+1])) == -1) {
				syslog(LLEV,"fwtkcfgerr: cannot map uid %s",av[x+1]);
				exit(1);
			}
			break;
		}

	for(x = 0; x < xecind; x++)
		if(!strcmp(av[x],"-group")) {
			if((rungid  = mapgid(av[x+1])) == -1) {
				syslog(LLEV,"fwtkcfgerr: cannot map gid %s",av[x+1]);
				exit(1);
			}
			break;
		}

	for(x = 0; x < xecind; x++)
		if(!strcmp(av[x],"-chroot")) {
			chrdir = av[x+1];
			break;
		}


	if(chrdir != (char *)0) {
		chdir("/");
		if((chdir(chrdir) != 0 || chroot(chrdir) != 0)) {
			syslog(LLEV,"fwtksyserr: cannot chroot %s: %m",chrdir);
			exit(1);
		}
		chdir("/");
	}

	if(rungid != -1 && setuid(rungid)) {
		syslog(LLEV,"fwtksyserr: cannot setgid %d: %m",rungid);
		exit(1);
	}

	if(runuid != -1 && setuid(runuid)) {
		syslog(LLEV,"fwtksyserr: cannot setuid %d: %m",runuid);
		exit(1);
	}

	syslog(LLEV,"permit host=%s/%s service=%s execute=%s",rhost,raddr,service,av[xecind + 1]);
	execv(av[xecind + 1],&av[xecind + 1]);
	syslog(LLEV,"fwtksyserr: exec: %s : %m",av[xecind + 1]);
	exit(1);
}
